src/app/util/ember-compatibility-functions.cpp

```diff
@@ -543,11 +580,6 @@ CHIP_ERROR ReadSingleClusterData(const SubjectDescriptor & aSubjectDescriptor, b
         attributeMetadata = emberAfLocateAttributeMetadata(aPath.mEndpointId, aPath.mClusterId, aPath.mAttributeId);
     }
 
-    if (attributeCluster == nullptr && attributeMetadata == nullptr)
```

Comment:
Per spec, the "no such attribute" error comes before ACL checks, so moving this block to after ACL checks is not OK.

Furthermore, returning values for FeatureMap and ClusterRevision if the _endpoint_ or _cluster_ does not exist is wrong.

We should probably set `attributeCluster` for these if we did not find metadata, leave this block be, and then after the ACL check if `attributeCluster` is set and our attr is FeatureMap or ClusterRevision we synthesize the value.

Alternately, we figure out how to not take this early return for these attributes if the cluster exists, then keep going to the normal code below, but if `emAfReadOrWriteAttribute` and the failure is "no attribute" then we know it's one of these attributes and we can synthesize the value there.  That would avoid duplicating all this "set up the response IB" code....

Reply:
See "8.4.3.2. Incoming Read Request and Subscribe Request Action Processing" in the spec.  Under "If the path is a concrete path" the checks for endpoint/cluster/attribute are done before the "requires a privilege" check.